src/api/controllers/db-controller.ts

```diff
@@ -828,43 +942,154 @@ export async function getTxFromDataStore(
-    }
-    const contractAbi: ClarityAbi = JSON.parse(contract.result.abi);
-    const functionAbi = contractAbi.functions.find(
-      fn => fn.name === parsedTx.contract_call.function_name
+    return parseContractCallMetadata(contract, parsedTx, dbTransaction);
+  }
+}
+
+function parseContractCallMetadata(
+  contract: FoundOrNot<DbSmartContract>,
+  parsedTx: ContractCallTransaction | MempoolContractCallTransaction,
+  dbTransaction: DbTx | DbMempoolTx
+) {
+  if (!contract.found) {
+    throw new Error(`Failed to lookup smart contract by ID ${parsedTx.contract_call.contract_id}`);
+  }
+  const contractAbi: ClarityAbi = JSON.parse(contract.result.abi);
+  const functionAbi = contractAbi.functions.find(
+    fn => fn.name === parsedTx.contract_call.function_name
+  );
+  if (!functionAbi) {
+    throw new Error(
+      `Could not find function name "${parsedTx.contract_call.function_name}" in ABI for ${parsedTx.contract_call.contract_id}`
     );
-    if (!functionAbi) {
-      throw new Error(
-        `Could not find function name "${parsedTx.contract_call.function_name}" in ABI for ${parsedTx.contract_call.contract_id}`
-      );
-    }
-    parsedTx.contract_call.function_signature = abiFunctionToString(functionAbi);
-    if (dbTx.contract_call_function_args) {
-      parsedTx.contract_call.function_args = readClarityValueArray(
-        dbTx.contract_call_function_args
-      ).map((c, fnArgIndex) => {
-        const functionArgAbi = functionAbi.args[fnArgIndex++];
-        return {
-          hex: bufferToHexPrefixString(serializeCV(c)),
-          repr: cvToString(c),
-          name: functionArgAbi.name,
-          type: getTypeString(functionArgAbi.type),
-        };
-      });
-    }
   }
+  parsedTx.contract_call.function_signature = abiFunctionToString(functionAbi);
+  if (dbTransaction.contract_call_function_args) {
+    parsedTx.contract_call.function_args = readClarityValueArray(
+      dbTransaction.contract_call_function_args
+    ).map((c, fnArgIndex) => {
+      const functionArgAbi = functionAbi.args[fnArgIndex++];
+      return {
+        hex: bufferToHexPrefixString(serializeCV(c)),
+        repr: cvToString(c),
+        name: functionArgAbi.name,
+        type: getTypeString(functionArgAbi.type),
+      };
+    });
+  }
+  return parsedTx;
+}
+
+export async function searchTxs(
+  db: DataStore,
+  args: GetTxsArgs | GetTxsWithEventsArgs
+): Promise<TransactionList> {
+  const minedTxs = await getTxsFromDataStore(db, args);
+
+  // filtering out mined transactions in canonical chain
+  const minedAndInCanonicalChain = minedTxs.filter(tx => tx.canonical && tx.microblock_canonical);
+
+  // filtering out tx_ids that are not in canonical chain
+  const nonCanonical = minedTxs.filter(tx => !tx.canonical && !tx.microblock_canonical);
+
+  // filtering out tx_ids that were not mined / found
+  const notMinedTransactions: string[] = args.txIds.filter(
+    txId => !minedTxs.find(minedTx => txId === minedTx.tx_id)
+  );
```

Comment:
Instead of traversing `minedTxs` twice and `args.txIds` once, you should do a `forEach` over `minedTxs` just once and populate all required temp arrays.

This function in particular (`searchTxs`) has an unusual number of filter/map/find calls that could be optimized into fewer traversals. I think some optimizations could be made.

Reply:
I've refactored for the other maps. However, this one is the only traversal I'm using for `args.txIds`. I wanted to get all the transaction ids that weren't mined.